src/app/beer_garden/local_plugins/manager.py

```diff
@@ -1,343 +1,519 @@
 # -*- coding: utf-8 -*-
-import logging
 
-from brewtils.models import Instance, System
+import json
+import logging
+import string
+from dataclasses import dataclass
+from importlib.machinery import SourceFileLoader
+from importlib.util import module_from_spec, spec_from_file_location
+from pathlib import Path
+from random import choice
+from time import sleep
+from types import ModuleType
+from typing import Dict
+from enum import Enum
+
+
+import sys
+from brewtils.specification import _SYSTEM_SPEC
+from brewtils.stoppable_thread import StoppableThread
 
 import beer_garden
-import beer_garden.db.api as db
-import beer_garden.queue.api as queue
-from beer_garden.errors import PluginStartupError
-from beer_garden.local_plugins.loader import LocalPluginLoader
-from beer_garden.local_plugins.plugin_runner import LocalPluginRunner
-from beer_garden.local_plugins.registry import LocalPluginRegistry
-from beer_garden.local_plugins.validator import LocalPluginValidator
-from beer_garden.queue.rabbit import get_routing_key
-
-
-class LocalPluginsManager(object):
-    """Manager that is capable of stopping/starting and restarting plugins"""
-
-    def __init__(self, shutdown_timeout):
-        self.logger = logging.getLogger(__name__)
-        self.loader = LocalPluginLoader.instance()
-        self.validator = LocalPluginValidator.instance()
-        self.registry = LocalPluginRegistry.instance()
-        self.shutdown_timeout = shutdown_timeout
-
-    def start_plugin(self, plugin):
-        """Start a specific plugin.
-
-        If a plugin cannot be found (i.e. it was not loaded) then it will not do anything
-        If the plugin is already running, it will do nothing.
-        If a plugin instance has not started after the PLUGIN_STARTUP_TIMEOUT it will
-        mark that instance as stopped
-
-        :param plugin: The plugin to start
-        :return: True if any plugin instances successfully started. False if the plugin
-        does not exist or all instances failed to start
+import beer_garden.config
+from beer_garden.errors import PluginValidationError
+from beer_garden.local_plugins.env_help import expand_string
+from beer_garden.local_plugins.runner import ProcessRunner
+
+CONFIG_NAME = "beer.conf"
+
+
+class ConfigKeys(Enum):
+    PLUGIN_ENTRY = 1
+    INSTANCES = 2
+    PLUGIN_ARGS = 3
+    ENVIRONMENT = 4
+    LOG_LEVEL = 5
+
+    NAME = 6
+    VERSION = 7
+    DESCRIPTION = 8
+    MAX_INSTANCES = 9
+    ICON_NAME = 10
+    DISPLAY_NAME = 11
+    METADATA = 12
+
+
+@dataclass
+class Plugin:
+    runner: ProcessRunner
+    instance_id: str = ""
+    restart: bool = False
+    dead: bool = False
```

Comment:
Unfortunately the place where this would be really useful is in the brewtils models, but dataclasses weren't added until Python 3.7. So we can only use them in beer-garden proper :cry: